src/Course/StateT.hs

```diff
@@ -267,32 +268,38 @@ instance Functor f => Functor (OptionalT f) where
 instance Monad f => Applicative (OptionalT f) where
   pure = OptionalT . pure . Full
 
-  -- (<*>) :: OptionalT (f (Optional (a -> b))) -> OptionalT (f (Optional a)) -> OptionalT (f (Optional b))
-  -- ffunc :: f (Optional (a -> b))
-  -- fa    :: f (Optional a)
-  -- func <*> :: f _ -> f b
-  OptionalT ffunc <*> OptionalT fa = OptionalT $ (<*>) ffunc <*> _
-
-  -- OptionalT ffunc <*> OptionalT fa = OptionalT $ do
-  --   maybeFunc <- ffunc
-  --   maybeA <- fa
-  --   case maybeFunc of
-  --     Empty -> Empty
-  --     Full func -> case maybeA of
-  --       Empty -> Empty
-  --       Full a -> Full $ func a
-
-
-  -- OptionalT ffunc <*> OptionalT fa = onFull t -> f (Optional a) Optional t
-  --   Applicative f => (t -> f (Optional a)) -> Optional t -> f (Optional a)
-  -- theirs: OptionalT $ (<*>) <$> ffunc <*> fa (but doesn't pass all tests?)
+  OptionalT ffunc <*> OptionalT fa = OptionalT $ do
+    maybeFunc <- ffunc
+    case maybeFunc of
+      Empty     -> pure Empty
+      Full func -> do
+        a <- fa
+        pure $ func <$> a
+    -- REVIEW:
+    -- Above, you suggested solutions like these over more point-free ones like theirs
+    -- (OptionalT f <*> OptionalT a = OptionalT (lift2 (<*>) f a)). Does this fall into the same
+    -- category? I'm still trying to build my intuition around when it makes sense to go point-free
+    -- and when it doesn't.
+    --
+    -- Also, their version fails tests 5-7
+    -- (https://github.com/maxrothman/fp-course/blob/e20fcd61979b76778836dd5277181937ad6facc3/test/Course/StateTTest.hs#L133)
+    -- Mine did too when I had `a <- fa` in the top do. Moving it to the bottom do fixed the issue.
+    --
+    -- Also not really sure what the suggestion to use onFull was about, they don't use it either.
+
+
 -- | Implement the `Monad` instance for `OptionalT f` given a Monad f.
 --
 -- >>> runOptionalT $ (\a -> OptionalT (Full (a+1) :. Full (a+2) :. Nil)) =<< OptionalT (Full 1 :. Empty :. Nil)
 -- [Full 2,Full 3,Empty]
 instance Monad f => Monad (OptionalT f) where
-  (=<<) =
-    error "todo: Course.StateT (=<<)#instance (OptionalT f)"
+  func =<< OptionalT fa = OptionalT $ do
+    maybeA <- fa
+    case maybeA of
+      Empty -> pure Empty
+      Full a -> runOptionalT . func $ a
+      -- REVIEW: un-OptionalT'ing this to just re-OptionalT it again seems dumb. There's no
```

Comment:
Well, that's the essence of what `=<<` does: it unwraps, applies the function, then rewraps. If you can do it in a clever way that doesn't actually involved unwrapping/rewrapping - more power to you, but in general this is indeed impossible. [Look at `MaybeT`](https://www.stackage.org/haddock/lts-13.6/transformers-0.5.5.0/src/Control-Monad-Trans-Maybe.html#MaybeT) for example.